src/debug/telemetry/telemetry.c

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2023 Intel Corporation. All rights reserved.
+//
+// Author: Tobiasz Dryjanski <tobiaszx.dryjanski@intel.com>
+
+#include <errno.h>
+#include <limits.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <adsp_debug_window.h>
+#include <mem_window.h>
+#include <zephyr/debug/sparse.h>
+#include <sof/debug/telemetry/telemetry.h>
+#include <sof/audio/module_adapter/module/generic.h>
+#include "adsp_debug_window.h"
+
+/* systic vars */
+int systick_counter;
+int prev_ccount;
+int perf_period_sum_;
+int perf_period_cnt_;
+struct perf_queue perf_queue = {0};
+
+static void perf_queue_append(struct perf_queue *q, size_t element)
+{
+	if (!q->full) {
+		q->elements[q->index] = element;
+		q->sum += element;
+		q->index++;
+		q->size++;
+		if (q->index >= AVG_PERF_MEAS_DEPTH) {
+			q->index = 0;
+			q->size = AVG_PERF_MEAS_DEPTH;
+			q->full = true;
+		}
+	} else {
+		/* no space, pop tail */
+		q->sum -= q->elements[q->index];
+		/* replace tail */
+		q->elements[q->index] = element;
+		q->sum += element;
+		/* move tail */
+		q->index++;
+		if (q->index >= AVG_PERF_MEAS_DEPTH)
+			q->index = 0;
+	}
+}
+
+static size_t perf_queue_avg(struct perf_queue *q)
+{
+	if (!q->size)
+		return 0;
+	return q->sum / q->size;
+}
+
+int telemetry_init(void)
+{
+	/* systick_init */
+	uint8_t slot_num = DW_TELEMETRY_SLOT;
+	volatile struct adsp_debug_window *window = ADSP_DW;
+	struct telemetry_wnd_data *wnd_data = (struct telemetry_wnd_data *)ADSP_DW->slots[slot_num];
+	struct system_tick_info *systick_info =
+			(struct system_tick_info *)wnd_data->system_tick_info;
+
+	window->descs[slot_num].type = ADSP_DW_SLOT_TELEMETRY;
+	window->descs[slot_num].resource_id = 0;
+	wnd_data->separator_1 = 0x0000C0DE;
```

Comment:
some magic number ?

Reply:
Separators were defined in reference FW. Those are only  used to identify telemetry structs when looking at the memory directly (e.g. using RWE), so those can be anything really.

Reply:
ok, so to be scalable (where FW can add and modify data payloads in a way that wont break host based tooling) we would need to put a small header as the separator to convey extra information instead of just a magic number e.g.
```c
struct telem_hdr {
    uint16_t magic; // 0xc0de
    uint16_t type; // type of C struct, i.e. what am I.
    uint16_t words;  // size of data elem in words
    uint16_t elem; // > 1 if payload is array
} __packed;
// payload follows here
```
This could convey all the info tooling needs and would allow FW to grow data and not break tooling along teh way. i.e. if tooling does not know about a ```type``` then it can go to the next header (and so on).

Reply:
The structure of telemetry data in memory window 2 here is the same as in reference firmware. If we want to be compatible it would be best to leave it as it is.

Reply:
We do need to maintain compatability but we also need to scale and grow as new features upstream. We can wrap any legacy/reference structure as-is, but they would need to be pre-fixed with above header so that new data could be added (without impacting any legacy data).

Reply:
Still not sure if we can just change the separators. Pre-fixing with the header would also move the fields from their designated places in the MW slot, and things would just stop working.

About scaling: memory window 2 is pretty scalable by itself. We can still extend the window to get more slots where we can add new structures.

Reply:
I think we have to live with the fixed location (slot + slot offset + separator) to keep compatibility.

Reply:
This is a debug mechanism that has to grow and scale as FW features, use cases and capabilities grow and scale (and they will), its way simpler and cheaper to fix the debug tooling (i.e. look for a header to find start, all other data is the same) than to continue with an ABI that cant grow with the FW.